designs/disruption-controls.md

```diff
@@ -0,0 +1,118 @@
+# Disruption Controls - (in v1beta1+)
+
+## Motivation
+Users have been asking for more control around the speed of Consolidation and the ability to restrict disruptions to their nodes. The demand for this is reflected by the +1s in [#1738](https://github.com/aws/karpenter/issues/1738). This is increasingly important as clusters can run at a highly varied scales, where some users have reported scale-down behaving too slow at higher scales, or too quick at smaller scales. Additionally, users that want to control when nodes can be disrupted must add PDBs to block node disruptions, add `do-not-evict` annotations on pods for sensitive workloads, or add `do-not-consolidate` annotations on nodes that shouldn't be consolidated. Natively integrating a better way to ensure nodes aren't disrupted will ease the burden that users have to take on with disruption.
+
+These controls will include a new `Disruption` block in the `v1beta1` `NodePool` that will contain all scale-down behavioral fields. Initially, this will include `Budgets`, which define a (1) parallelism of how many nodes can be deprovisioned at a time and a (2) cron schedule that determines when the parallelism applies, and a `ConsolidateAfter` field that will allow users to affect the speed at which Karpenter will scale down underutilized nodes. Future scale-down behavioral fields should be colocated with these fields.
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized || WhenEmpty
```

Comment:
Did you consider moving this into `budgets` so empty nodes can be deleted after a while during business hours, and full consolidation is done the rest of the time?

Reply:
We're using this setting to drive what type of cost-consolidation method's can be done for nodes owned by this provisioner. We've heard that users may want to set `actions` that determine which types of disruption are restricted, which I think is what you're asking for. We're opting to explore that as a follow-up to this.

Reply:
I think the naming of `consolidationPolicy` and  `WhenUnderutilized`  is kind of poor. The reason is node pools doesn't talk about underutilization and through out the docs consolidation is talked about as a method of deprovisioning similar to emptiness and drift and so on. 

I also don't fully understand this new policy since there are 5 different deprovisioning mechanisms but this only touches 2. 

Could the control be more like:
```
disruption:
  deprovisionWhenEmpty: enabled || disabled
  deprovisionWhenExpired: enabled || disabled
  deprovisionedWhenConsolidated: enabled || disabled
  # interruption + drift
```
or 

```
disruption:
  emptiness: 
    enabled: true
    deprovisionAfter: 10m || never
  # the other deprovision methods
```

Reply:
We haven't made the changes to the docs yet, as this design relates to v1beta1 APIs, and the v1beta1 APIs haven't been released. We'll make the doc changes before those are released. We'll definitely reorganize the docs to match what we're talking about here. 

The idea is that underutilization encapsulates the current emptiness and consolidation, but consolidation also does emptiness checks. Calling it `consolidationPolicy` will allow users to choose how they want their nodes Consolidated, mapping to the two existing v1alpha5 APIs. `WhenUnderutilized` is a more general way to say "if it saves on cost" whereas `WhenEmpty` is very specific in that we'll only delete empty nodes. 

We're also trying to move away from the `deprovision` language, as we only used that because the API was called the Provisioner, where "deprovisioning" is the opposite. Using `disruption` is more akin to Kubernetes concepts (e.g. PDBs, [other disruptions](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/))

Some fields in the past had the semantic of omitting the field meant disabled, but that meant we couldn't persist defaults we wanted, requiring another field for toggling. Using the `10m || Never` helps us do all of these with one field. The idea is that a new user for Karpenter shouldn't have to think about these concepts and should be able to rely on Karpenter picking sane defaults, but allow users who don't want them to disable/modify them.

Reply:
Make sense thanks